File: test/python/test_approximate_equivalence.py
# Copyright (c) 2023 - 2026 Chair for Design Automation, TUM
# Copyright (c) 2025 - 2026 Munich Quantum Software Company GmbH
# All rights reserved.
#
# SPDX-License-Identifier: MIT
#
# Licensed under the MIT License

"""Tests for approximate equivalence checking."""

from __future__ import annotations

import math

import pytest
from qiskit import AncillaRegister, QuantumCircuit
from qiskit.circuit import Parameter

from mqt import qcec
from mqt.qcec.pyqcec import Configuration, EquivalenceCriterion


@pytest.fixture
def phase_pair() -> tuple[QuantumCircuit, QuantumCircuit]:
    """Return circuits with projective Hilbert--Schmidt distance one half."""
    identity = QuantumCircuit(1)
    phase = QuantumCircuit(1)
    phase.p(math.pi / 3, 0)
    return identity, phase


def test_configuration_options_accept_approximate_pair(
    phase_pair: tuple[QuantumCircuit, QuantumCircuit],
) -> None:
    """Configure approximate equivalence through a Configuration object."""
    identity, phase = phase_pair
    config = Configuration()
    config.execution.parallel = False
    config.functionality.check_approximate_equivalence = True
    config.functionality.approximate_checking_threshold = 0.51

    result = qcec.verify(identity, phase, configuration=config)

    assert result.equivalence == EquivalenceCriterion.equivalent
    assert result.started_simulations == 0


def test_keyword_options_reject_pair_beyond_threshold(
    phase_pair: tuple[QuantumCircuit, QuantumCircuit],
) -> None:
    """Configure approximate equivalence through verify keyword arguments."""
    identity, phase = phase_pair

    result = qcec.verify(
        identity,
        phase,
        parallel=False,
        check_approximate_equivalence=True,
        approximate_checking_threshold=0.49,
    )

    assert result.equivalence == EquivalenceCriterion.not_equivalent
    assert result.started_simulations == 0


@pytest.mark.parametrize("parallel", [False, True])
def test_default_simulation_is_not_used_for_process_distance(parallel: bool) -> None:
    """Do not let state simulation overrule the configured process metric."""
    identity = QuantumCircuit(1)
    x_gate = QuantumCircuit(1)
    x_gate.x(0)

    result = qcec.verify(
        identity,
        x_gate,
        parallel=parallel,
        check_approximate_equivalence=True,
        approximate_checking_threshold=1.0,
    )

    assert result.equivalence == EquivalenceCriterion.equivalent
    assert result.started_simulations == 0


@pytest.mark.parametrize("threshold", [-0.1, 1.1, math.nan, math.inf, -math.inf])
def test_invalid_threshold_is_rejected(threshold: float) -> None:
    """Require a finite process-distance threshold in the unit interval."""
    circuit = QuantumCircuit(1)

    with pytest.raises(ValueError, match="threshold"):
        qcec.verify(
            circuit,
            circuit,
            check_approximate_equivalence=True,
            approximate_checking_threshold=threshold,
        )


def test_partial_approximate_equivalence_is_rejected() -> None:
    """Reject the undefined combination of partial and approximate checking."""
    circuit = QuantumCircuit(1)

    with pytest.raises(ValueError, match="partial"):
        qcec.verify(
            circuit,
            circuit,
            check_partial_equivalence=True,
            check_approximate_equivalence=True,
        )


def test_approximate_equivalence_with_ancilla_is_rejected() -> None:
    """Reject process-distance checking for non-unitary ancilla semantics."""
    ancilla = AncillaRegister(1)
    circuit = QuantumCircuit(1)
    circuit.add_register(ancilla)
    circuit.h(ancilla[0])

    with pytest.raises(ValueError, match="ancilla"):
        qcec.verify(circuit, circuit, check_approximate_equivalence=True)


def test_approximate_equivalence_for_symbolic_circuits_is_rejected() -> None:
    """Reject approximate checking without a universal symbolic definition."""
    theta = Parameter("theta")
    symbolic = QuantumCircuit(1)
    symbolic.rz(theta, 0)

    with pytest.raises(ValueError, match="symbolic"):
        qcec.verify(symbolic, symbolic, check_approximate_equivalence=True)


def test_approximate_equivalence_requires_a_supported_dd_checker() -> None:
    """Require a checker that implements the configured process metric."""
    circuit = QuantumCircuit(1)

    with pytest.raises(ValueError, match=r"alternating.*construction"):
        qcec.verify(
            circuit,
            circuit,
            check_approximate_equivalence=True,
            run_alternating_checker=False,
            run_construction_checker=False,
        )


def test_approximate_options_are_serialized() -> None:
    """Expose approximate-equivalence options in the configuration JSON."""
    config = Configuration()
    config.functionality.check_approximate_equivalence = True
    config.functionality.approximate_checking_threshold = 0.25

    functionality = config.json()["functionality"]

    assert functionality["check_approximate_equivalence"] is True
    assert functionality["approximate_checking_threshold"] == pytest.approx(0.25)

File: test/test_approximate_equivalence.cpp
/*
 * Copyright (c) 2023 - 2026 Chair for Design Automation, TUM
 * Copyright (c) 2025 - 2026 Munich Quantum Software Company GmbH
 * All rights reserved.
 *
 * SPDX-License-Identifier: MIT
 *
 * Licensed under the MIT License
 */

#include "Configuration.hpp"
#include "EquivalenceCheckingManager.hpp"
#include "EquivalenceCriterion.hpp"
#include "dd/DDDefinitions.hpp"
#include "ir/QuantumComputation.hpp"
#include "ir/operations/Expression.hpp"

#include <cstdint>
#include <gtest/gtest.h>
#include <limits>
#include <stdexcept>
#include <string>

namespace {

enum class DDChecker : std::uint8_t { Alternating, Construction };

std::string checkerName(const testing::TestParamInfo<DDChecker>& info) {
  return info.param == DDChecker::Alternating ? "Alternating" : "Construction";
}

class ApproximateEquivalenceTest : public testing::TestWithParam<DDChecker> {
protected:
  [[nodiscard]] static ec::Configuration configuration(const double threshold) {
    ec::Configuration config{};
    config.execution.parallel = false;
    config.execution.runSimulationChecker = false;
    config.execution.runAlternatingChecker =
        GetParam() == DDChecker::Alternating;
    config.execution.runConstructionChecker =
        GetParam() == DDChecker::Construction;
    config.execution.runZXChecker = false;
    config.functionality.checkApproximateEquivalence = true;
    config.functionality.approximateCheckingThreshold = threshold;
    return config;
  }
};

TEST_P(ApproximateEquivalenceTest, AcceptsDistanceBelowThreshold) {
  auto qc1 = qc::QuantumComputation(1);
  auto qc2 = qc::QuantumComputation(1);
  qc1.x(0);
  qc2.x(0);
  qc2.ry(dd::PI / 3., 0);

  // D_HS(X, RY(pi/3) X) = sin(pi/6) = 0.5.
  auto manager = ec::EquivalenceCheckingManager(qc1, qc2, configuration(0.51));
  manager.run();

  EXPECT_EQ(manager.equivalence(), ec::EquivalenceCriterion::Equivalent);
}

TEST_P(ApproximateEquivalenceTest, RejectsDistanceAboveThreshold) {
  auto qc1 = qc::QuantumComputation(1);
  auto qc2 = qc::QuantumComputation(1);
  qc1.x(0);
  qc2.x(0);
  qc2.ry(dd::PI / 3., 0);

  // D_HS(X, RY(pi/3) X) = sin(pi/6) = 0.5.
  auto manager = ec::EquivalenceCheckingManager(qc1, qc2, configuration(0.49));
  manager.run();

  EXPECT_EQ(manager.equivalence(), ec::EquivalenceCriterion::NotEquivalent);
}

TEST_P(ApproximateEquivalenceTest, IncludesMaximumDistanceBoundary) {
  auto qc1 = qc::QuantumComputation(1);
  auto qc2 = qc::QuantumComputation(1);
  qc1.x(0);
  qc2.z(0);

  // D_HS(X, Z) = 1, which is the inclusive upper threshold boundary.
  auto manager = ec::EquivalenceCheckingManager(qc1, qc2, configuration(1.));
  manager.run();

  EXPECT_EQ(manager.equivalence(), ec::EquivalenceCriterion::Equivalent);
}

TEST_P(ApproximateEquivalenceTest, IncludesExactZeroBoundary) {
  auto qc1 = qc::QuantumComputation(1);
  auto qc2 = qc::QuantumComputation(1);
  qc1.h(0);
  qc2.h(0);

  auto manager = ec::EquivalenceCheckingManager(qc1, qc2, configuration(0.));
  manager.run();

  EXPECT_EQ(manager.equivalence(), ec::EquivalenceCriterion::Equivalent);
}

TEST_P(ApproximateEquivalenceTest, PreservesExactGlobalPhaseResult) {
  auto qc1 = qc::QuantumComputation(1);
  auto qc2 = qc::QuantumComputation(1);
  qc1.x(0);
  qc2.x(0);
  qc2.z(0);
  qc2.x(0);
  qc2.z(0);
  qc2.x(0);

  auto manager = ec::EquivalenceCheckingManager(qc1, qc2, configuration(0.));
  manager.run();

  EXPECT_EQ(manager.equivalence(),
            ec::EquivalenceCriterion::EquivalentUpToGlobalPhase);
}

INSTANTIATE_TEST_SUITE_P(DecisionDiagramCheckers, ApproximateEquivalenceTest,
                         testing::Values(DDChecker::Alternating,
                                         DDChecker::Construction),
                         checkerName);

class ApproximateConfigurationTest : public testing::Test {
protected:
  static ec::Configuration validConfiguration() {
    ec::Configuration config{};
    config.execution.parallel = false;
    config.execution.runSimulationChecker = false;
    config.execution.runAlternatingChecker = true;
    config.execution.runConstructionChecker = false;
    config.execution.runZXChecker = false;
    config.functionality.checkApproximateEquivalence = true;
    return config;
  }

  qc::QuantumComputation qc1{2};
  qc::QuantumComputation qc2{2};
};

class InvalidApproximateThresholdTest
    : public ApproximateConfigurationTest,
      public testing::WithParamInterface<double> {};

TEST_P(InvalidApproximateThresholdTest, RejectsInvalidThreshold) {
  qc1.x(0);
  qc2.x(0);
  auto config = validConfiguration();
  config.functionality.approximateCheckingThreshold = GetParam();
  auto manager = ec::EquivalenceCheckingManager(qc1, qc2, config);

  EXPECT_THROW(manager.run(), std::invalid_argument);
}

INSTANTIATE_TEST_SUITE_P(
    InvalidThresholds, InvalidApproximateThresholdTest,
    testing::Values(-0.01, 1.01, std::numeric_limits<double>::quiet_NaN(),
                    std::numeric_limits<double>::infinity(),
                    -std::numeric_limits<double>::infinity()));

TEST_F(ApproximateConfigurationTest, RejectsPartialEquivalenceCombination) {
  qc1.x(0);
  qc2.x(0);
  auto config = validConfiguration();
  config.functionality.checkPartialEquivalence = true;
  auto manager = ec::EquivalenceCheckingManager(qc1, qc2, config);

  EXPECT_THROW(manager.run(), std::invalid_argument);
}

TEST_F(ApproximateConfigurationTest, RejectsActedUponAncillaryQubits) {
  qc1.x(0);
  qc1.x(1);
  qc2.x(0);
  qc2.x(1);
  qc1.setLogicalQubitAncillary(1);
  qc2.setLogicalQubitAncillary(1);
  auto manager = ec::EquivalenceCheckingManager(qc1, qc2, validConfiguration());

  EXPECT_THROW(manager.run(), std::invalid_argument);
}

TEST_F(ApproximateConfigurationTest, RejectsActedUponGarbageQubits) {
  qc1.x(0);
  qc1.x(1);
  qc2.x(0);
  qc2.x(1);
  qc1.setLogicalQubitGarbage(1);
  qc2.setLogicalQubitGarbage(1);
  auto manager = ec::EquivalenceCheckingManager(qc1, qc2, validConfiguration());

  EXPECT_THROW(manager.run(), std::invalid_argument);
}

TEST_F(ApproximateConfigurationTest, RejectsSymbolicCircuits) {
  const auto variable = sym::Variable("theta");
  const auto angle = qc::Symbolic{sym::Term<dd::fp>{variable}};
  qc1.rx(angle, 0);
  qc2.rx(angle, 0);
  auto manager = ec::EquivalenceCheckingManager(qc1, qc2, validConfiguration());

  EXPECT_THROW(manager.run(), std::invalid_argument);
}

TEST_F(ApproximateConfigurationTest, RequiresDecisionDiagramChecker) {
  qc1.x(0);
  qc2.x(0);
  auto config = validConfiguration();
  config.execution.runAlternatingChecker = false;
  config.execution.runSimulationChecker = true;
  config.execution.runZXChecker = true;
  auto manager = ec::EquivalenceCheckingManager(qc1, qc2, config);

  EXPECT_THROW(manager.run(), std::invalid_argument);
}

TEST_F(ApproximateConfigurationTest, RejectsConfigurationWithoutCheckers) {
  qc1.x(0);
  qc2.x(0);
  auto config = validConfiguration();
  config.execution.runAlternatingChecker = false;
  auto manager = ec::EquivalenceCheckingManager(qc1, qc2, config);

  EXPECT_THROW(manager.run(), std::invalid_argument);
}

TEST_F(ApproximateConfigurationTest, ValidatesConfigurationChangesOnRun) {
  qc1.x(0);
  qc2.x(0);
  auto config = validConfiguration();
  config.functionality.checkApproximateEquivalence = false;
  auto manager = ec::EquivalenceCheckingManager(qc1, qc2, config);
  manager.getConfiguration().functionality.checkApproximateEquivalence = true;
  manager.getConfiguration().functionality.approximateCheckingThreshold =
      std::numeric_limits<double>::quiet_NaN();

  EXPECT_THROW(manager.run(), std::invalid_argument);
}

class ApproximateDefaultFlowTest : public testing::TestWithParam<bool> {};

TEST_P(ApproximateDefaultFlowTest, DisablesUnsupportedCheckers) {
  auto qc1 = qc::QuantumComputation(1);
  auto qc2 = qc::QuantumComputation(1);
  qc1.x(0);
  qc2.x(0);
  qc2.ry(dd::PI / 3., 0);

  auto config = ec::Configuration{};
  config.execution.parallel = GetParam();
  config.execution.nthreads = 2U;
  // Exercise the actual parallel flow with both supported DD checkers.
  config.execution.runConstructionChecker = GetParam();
  config.functionality.checkApproximateEquivalence = true;
  config.functionality.approximateCheckingThreshold = 0.51;
  ASSERT_TRUE(config.execution.runSimulationChecker);
  ASSERT_TRUE(config.execution.runZXChecker);
  auto manager = ec::EquivalenceCheckingManager(qc1, qc2, config);

  manager.run();

  EXPECT_EQ(manager.equivalence(), ec::EquivalenceCriterion::Equivalent);
  EXPECT_FALSE(manager.getConfiguration().execution.runSimulationChecker);
  EXPECT_FALSE(manager.getConfiguration().execution.runZXChecker);
  EXPECT_EQ(manager.getResults().startedSimulations, 0U);
  EXPECT_EQ(manager.getResults().performedSimulations, 0U);
}

INSTANTIATE_TEST_SUITE_P(SequentialAndParallel, ApproximateDefaultFlowTest,
                         testing::Bool());

TEST(ApproximateConfigurationJsonTest, SerializesApproximateOptions) {
  auto config = ec::Configuration{};
  config.functionality.checkApproximateEquivalence = true;
  config.functionality.approximateCheckingThreshold = 0.125;

  const auto json = config.json();

  EXPECT_TRUE(json["functionality"]["check_approximate_equivalence"]);
  EXPECT_DOUBLE_EQ(json["functionality"]["approximate_checking_threshold"],
                   0.125);
}

} // namespace
